CAN-FD Support

davidlenfesty opened the pull request:
I basically just copied over the old CAN implementation and tweaked it to use a 64-byte MTU and deal with the weird DLC values and padding.

Not ship-shape in it's current state just wanted to start work on it.

Probably needs to be re-worked a large amount.

davidlenfesty committed:
transport/can: migrate legacy CAN to new module.

In preparation for CAN FD support

davidlenfesty committed:
transport/can: clone legacy CAN into FD

davidlenfesty committed:
transport/can/fd: first attempt at making it work

davidlenfesty committed:
transport/can/fd: update some comments

davidlenfesty commented:
Took a look at STM appnote 5348 to confirm my understanding of the differences between CAN 2.0 and FD, and it seems the *only* thing that matters to this implementation are semantics around DLC, which were accounted for in the third commit.

There is some room to merge CAN 2.0 and CAN FD, but to do that in a compile-time way requires more progress in const generics from my understanding, (or macros, but eh) so I want to leave that until there's better support in Rust for that sort of thing.

This also requires testing, I haven't tested it at all. @teamplayer3, you may want to do the testing, as you said it would be useful for your thesis :). Otherwise I'll get around to testing it when I get around to it.

davidlenfesty commented:
It also affects some other PRs so even if it gets approved I'll wait until those get merged in so everything works cleanly.

teamplayer3 commented:
For testing, it would be great to merge `alex/no_std-session` into the main branch and then rebase this branch on the new state. This would allow using the `embedded` example to test the CANFD implementation.

teamplayer3 commented on uavcan/src/transport/can/mod.rs:
```diff
@@ -1,336 +1,15 @@
-    toggle: bool,
-    is_start: bool,
-}
+pub use fd::*;
```
I wouldn't remove the namespaces from the imports. I think importing the different structures like `transport::can::fd::CanFd` and `transport::can::legacy::Can` should be fine. Or do you think this is better if the two implementations will be merged in the future?

davidlenfesty committed:
transport/can: update legacy and fd for various changes

davidlenfesty committed:
rustfmt

davidlenfesty committed:
rustfmt you have failed me

davidlenfesty committed:
transport/can: satisfy clippy

davidlenfesty committed:
transport/can: remove extra namespace crunching

it's a bit hacky and will likely change if Can 2.0 and fd support are
ever properly merged together.
